testsuite/classlibrary/TestServer_clientID_booted.sc

```diff
@@ -70,4 +70,128 @@ TestServer_clientID_booted : UnitTest {
+
+//		1.wait;
+
+		// now login again from the same server object
+		// should get "This seems to be a login after loss of network ... all is well" post.
+		server.startAliveThread;
+
+		cond = Condition.new;
+
+		timer = fork { 3.wait; cond.unhang };
+		server.doWhenBooted {
+			"% - server login timed out.\n".postf(thisMethod);
+			timer.stop;
+			cond.unhang;
+		};
+
+		cond.hang;
+
+		this.assert(server.serverRunning,
+			"after repeated login, server should be known to be fully running."
+		);
+
+		this.assert(server.clientID == 3,
+			"after repeated login, remote client should get the same clientID from server process."
+		);
+
+		this.assert(server.nextNodeID > lastNodeID,
+			"after repeated login, nodeID allocator should not be reset."
+		);
+
+		server.quit;
+		server.remove;
+
+	}
+
+	// test that logins to a (pseudo-) remote server process work as intended
+	// first login initializes fully,
+	// second login from same address gets the same clientID, and sets running state.
+	test_remoteLogins {
+		var options, serverPid, remote1, remote2, cond, timer;
+
+		options = ServerOptions.new;
+		options.maxLogins = 4;
+
+		// simulate a remote server process by starting a server process independently of SC
+		serverPid = unixCmd(Server.program + options.asOptionsString);
+
+		1.wait;
+
+		// login with standard Server.remote method to test for
+		remote1 = Server.remote(\remoteLogin1, options: options, clientID: 3);
+		cond = Condition.new;
+		timer = fork { 3.wait; cond.unhang };
+
+		remote1.doWhenBooted {
+			"% - server first login timed out.\n".postf(thisMethod);
+			timer.stop;
+			cond.unhang;
+		};
+		cond.hang;
+
+		this.assert(remote1.serverRunning,
+			"after first login, remote client should be known to be fully running."
+		);
+
+		this.assert(remote1.clientID == 3,
+			"after first login, remote client should have its requested clientID."
+		);
+
+		// now login again from a different server object,
+		// but from the same client address (in the test, local, in reality, it would be remote).
+		// This simulates a crashed remote client which gets restarted,
+		// and logs in with its old setup data.
+
+		// Because client netaddr is already registered in the server process,
+		// the response goes thru prHandleNotifyFailString, and prHandleLoginWhenAlreadyRegistered
+
+		remote2 = Server.remote(\remoteLogin2, options: options);
+		cond = Condition.new;
+		timer = fork { 3.wait; cond.unhang };
+
+		remote2.doWhenBooted {
+			"% - server repeated login timed out.\n".postf(thisMethod);
+			timer.stop;
+			cond.unhang;
+		};
+
+		cond.hang;
+
+		this.assert(remote2.serverRunning,
+			"after re-login from a new server object, remote client should be known to be fully running."
+		);
+
+		this.assert(remote2.clientID == 3,
+			"after re-login from a new server object, remote client should have its requested clientID."
+		);
+
+		// cleanup
+		remote1.remove;
+		remote2.remove;
+		unixCmd("kill -9" + serverPid);
```

Comment:
this is not portable at all

Reply:
OK. Is there a crossplatform version of kill?
or is it OK to use Server.killAll in the testsuite?
I would have preferred that, but assumed it is cleaner to only kill the exact process and nothing else.

Reply:
> Is there a crossplatform version of kill?

Windows uses taskkill.

> is it OK to use Server.killAll in the testsuite?

No. The single process approach is much more sound. Since we have `Platform.killAll` already, it would probably not be such a bad idea to add `killProcess` that accepts a pid. Just a suggestion. That should be done in a separate PR.

Reply:
ok

Reply:
I have near zero windows shell skills - would the methods below work?
in UnixPlatform:
	kill { |pid|
		("kill -9 " ++ pid).unixCmd;
	}
and in WindowsPlatform:
	kill { |pid|
		("taskkill /F /pid " ++ pid).unixCmd;
	}

Reply:
I think so.. I can't test at the moment